File: 43/src/main/scala/Problem43.scala
object Problem43 {

  lazy val answer: Long =
    (0 to 9).permutations.map(_.mkString)
      .filter({ s =>
        s.substring(1, 4).toInt % 2 == 0 &&
        s.substring(2, 5).toInt % 3 == 0 &&
        s.substring(3, 6).toInt % 5 == 0 &&
        s.substring(4, 7).toInt % 7 == 0 &&
        s.substring(5, 8).toInt % 11 == 0 &&
        s.substring(6, 9).toInt % 13 == 0 &&
        s.substring(7, 10).toInt % 17 == 0
      })
      .map(_.toLong)
      .sum

}

File: haskell/src/Euler/Problems/Problem43.hs
module Euler.Problems.Problem43
  ( answer
  , substrings
  , predicate
  ) where

import Euler.Prelude

import Data.List (sum, and, filter, zipWith, drop)

import Euler.Util.Arithmetic (divides)
import Euler.Util.Digit (digits, unDigits)
import Euler.Util.List (sliding)
import Euler.Util.Pandigital (pandigitals)
import Euler.Util.Prime (primes)

answer :: Integer
answer =
  sum $ filter predicate pandigitals

predicate :: Integer -> Bool
predicate i =
  and $ zipWith divides (primes :: [Integer]) (substrings i)

{- |

Three-digit substrings of the decimal representation of the input.

>>> substrings 1406357289
[406,63,635,357,572,728,289]

-}
substrings :: Integral a => a -> [a]
substrings =
  digits 10 >>>
  sliding 3 >>>
  drop 1 >>>
  fmap (unDigits (10 :: Integer))
